### PR TITLE
trails_handler supports parameter

### PR DESCRIPTION
**Pull request by turtleDeng**
cowboy_trails is a great project.
trails_handler support parameters will be a very good feature.
```
https://github.com/inaka/cowboy_swagger/blob/master/src/cowboy_swagger_json_handler.erl#L31
```

### CODE REVIEW

**Comment by elbrujohalcon** [comment]
Not a bad idea, but I guess it requires many more changes than the ones you just submitted, right?

**Comment by turtleDeng** [comment]
This is a compatibility modification
Support the use
```
trails:trails([{cowboy_swagger_handler, #{server => xxx}}]) and
trails:trails([cowboy_swagger_handler])
```

**Comment by turtleDeng** [comment]
Our project will listen to multiple http ports, and their APIs are different, trails: trails does not support parameters will cause me to be unable to open swaggerUI

**Comment by elbrujohalcon** [comment]
Sorry, @turtleDeng … But you'll have to be a bit more detailed than that.
On one hand, I kinda see what you're aiming to achieve… but on the other hand your PR seems to imply that the rest of the code in this project already supports that and I have no compelling reason to think that's true.

**Comment by elbrujohalcon** [comment]
> Our project will listen to multiple http ports, and their APIs are different, trails: trails does not support parameters will cause me to be unable to open swaggerUI

I see… and just by changing that single line in `trails_handler` you make it work? That's what seems odd to me. Don't you need to change other stuff in the project to add support for it?

**Comment by elbrujohalcon** [comment]
You, at the very least, need to add a new callback to `trails_handler`. And probably make it optional, I think.

**Comment by turtleDeng** [comment]
Yes, add a trails/1 callback

**Comment by elbrujohalcon** [comment]
Great! Now, can you please do the following things?
- [x] make the new callback optional (as a matter of fact, you'll probably like both callbacks to be optional now… so you can implement just one of them).
- [x] add some tests to `trails_SUITE` for this new feature
- [x] adjust the documentation for `trails_handler` callbacks in the README

File: test/trails_test3_handler.erl [hunk 1]
```diff
@@ -0,0 +1,15 @@
+-module(trails_test3_handler).
```
Comment (elbrujohalcon) [inline test/trails_test3_handler.erl:1]: For this module to be tested, you need to add it to `trails_SUITE:basic_trails_routes/1`… or create a new test function for it.

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

File: test/trails_test2_handler.erl [hunk 2]
```diff
@@ -3,10 +3,12 @@
...
--export([trails/0]).
+-export([trails/0, trails/1]).
 
 trails() ->
   [
    {"/api/resource3/[:id]", trails_test2_handler, []},
    {"/api/:id/resource4", trails_test2_handler, [arg0]}
   ].
+
+trails(_Opts) -> [].
```
Comment (elbrujohalcon) [inline test/trails_test2_handler.erl]: Now… if you make this callback optional, you won't need to add this boilerplate here.

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

File: test/trails_SUITE.erl [hunk 3]
```diff
@@ -309,17 +309,25 @@ basic_trails_routes(_Config) ->
   ExpectedResponse3 = StaticRoutes ++
     [ {"/api/resource3/[:id]", trails_test2_handler, []}
     , {"/api/:id/resource4", trails_test2_handler, [arg0]}
-    , {"/api/resource1/[:id]", trails_test_handler, []}
+    , {"/api/resource1/[:id]", trails_test_handler  , []}
```
Comment (elbrujohalcon) [inline test/trails_SUITE.erl]:
```suggestion
    , {"/api/resource1/[:id]", trails_test_handler, []}
```

File: src/trails_handler.erl [hunk 4]
```diff
@@ -8,6 +8,8 @@
 
 %% @doc Returns the cowboy routes defined in the called module.
 -callback trails() -> trails:trails().
+-callback trails(Opts :: map()) -> trails:trails().
```
Comment (elbrujohalcon) [inline src/trails_handler.erl:11]:
```suggestion
-callback trails(Opts :: map()) -> trails:trails().

-optional_callbacks([trails/0, trails/1]).
```

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
Two additional small changes and we're done.

File: src/trails_handler.erl (hunk 4 above)
Comment (turtleDeng) [reply src/trails_handler.erl:11]: Sorry, I am busy, it has been modified

**Review by elbrujohalcon — APPROVED** [review]

**Comment by elbrujohalcon** [comment]
Well… now we have some dialyzer issue in our hands… 🤔

**Pull request merged**